### PR TITLE
asserts,many: start supporting structured headers using the new parseHeaders

### PR DESCRIPTION
**Pull request by pedronis**
This starts supporting the new structured headers using the new parseHeaders.

It also switches most handling of assertion headers from map[string]string to map[string]interface{} (except when querying for now).

Notice that primary keys are expected to stay/must be strings.

It fixes and adapts tests.

### CODE REVIEW

**Comment by chipaca** [comment]
wow, quite the branch. Glad it's you and not me doing this :-) :+1:

File: asserts/asserts_test.go [hunk 1]
```diff
@@ -518,3 +521,20 @@ func (as *assertsSuite) TestAssembleRoundtrip(c *C) {
...
+func (as *assertsSuite) TestAssembleHeadersCheck(c *C) {
+	encoded := []byte("type: test-only\n" +
+		"authority-id: auth-id2\n" +
+		"primary-key: abc\n" +
+		"revision: 5" +
+		"\n\n" +
+		"openpgp c2ln")
+	a, err := asserts.Decode(encoded)
+	c.Assert(err, IsNil)
+	cont, sig := a.Signature()
```
Comment (emgee) [inline asserts/asserts_test.go]: afaict, the block above is really just to get valid content and signature for the actual test below. maybe add a comment above the block to make it clear it's just helper code?

File: asserts/database.go [hunk 2]
```diff
@@ -293,7 +293,7 @@ func (db *Database) Add(assert Assertion) error {
 
 	keyValues := make([]string, len(assertType.PrimaryKey))
 	for i, k := range assertType.PrimaryKey {
-		keyVal := assert.Header(k)
+		keyVal := assert.HeaderString(k)
 		if keyVal == "" {
 			return fmt.Errorf("missing primary key header: %v", k)
```
Comment (emgee) [inline asserts/database.go]: this error message could probably be improved a tiny bit to say "missing or non-string primary key header: %v". might help with debugging.

File: asserts/headers.go [hunk 3]
```diff
@@ -162,3 +171,92 @@ func parseList(first int, lines []string, baseIndent int) (value interface{}, fi
...
+		if len(x) == 0 {
+			return // simply omit
+		}
+		buf.WriteByte('\n')
+		buf.WriteString(intro)
+		pfx := nestingPrefix(baseIndent, listPrefix)
+		for _, elem := range x {
+			appendEntry(buf, pfx, elem, baseIndent+len(listPrefix)-1)
+		}
+	}
```
Comment (emgee) [inline asserts/headers.go]: probably worth having a `default` that panics here too.

**Comment by emgee** [comment]
+1 with a few small suggestions.

**Comment by pedronis** [comment]
retest this please

**Pull request merged**